clientconn.go

```diff
@@ -1621,3 +1626,62 @@ func (cc *ClientConn) connectionError() error {
 	defer cc.lceMu.Unlock()
 	return cc.lastConnectionError
 }
+
+func (cc *ClientConn) parseTargetAndFindResolver() (resolver.Builder, error) {
+	channelz.Infof(logger, cc.channelzID, "original dial target is: %q", cc.target)
+	parsedTarget, err := parseTarget(cc.target)
+	if err != nil {
+		channelz.Infof(logger, cc.channelzID, "dial target %q parse failed: %v", cc.target, err)
+	} else {
+		channelz.Infof(logger, cc.channelzID, "parsed dial target is: %+v", parsedTarget)
+	}
+	rb := cc.getResolver(parsedTarget.Scheme)
+	if rb != nil {
+		cc.parsedTarget = parsedTarget
+		return rb, nil
+	}
+
+	// We are here because the user's dial target did not contain a scheme or
+	// specified an unregistered scheme. We should fallback to the default
+	// scheme, except when no scheme was specified in the user's dial target and
+	// a custom dialer was specified. In the latter case, we should always use
+	// passthrough scheme.
+	defScheme := resolver.GetDefaultScheme()
+	if parsedTarget.Scheme == "" && cc.dopts.copts.Dialer != nil {
+		defScheme = "passthrough"
+	}
```

Comment:
Hmmmm, is this okay?  It might be.  It might not.

If the user has a custom dialer and they use `localhost:<port>` or `hostname:<port>`, we might want to still use `passthrough` as `defScheme`.  What would happen if we change this to just `if Dialer != nil {`?

Reply:
> where we might want this to probably get a `dns` scheme when we make that the default

True, but if there is a custom dialer we should pass the target directly to the user's dialer to keep existing behavior.  With a new `NewClient` method, we don't have to worry about backward compatibility, and we can always make DNS the default and maybe even pretend like passthrough doesn't exist / is invalid.